### DISCUSSION TITLE
aerich is too old that it struggles to advance its popularity

### DISCUSSION THREAD

**Discussion by licongy** [tortoise/tortoise-orm#1762] [Q&A] [answered]
There were so many bugs and problems with Aerich that the development experience with tortoise was so bad. I had to spend more time going back to studying SQL syntax to modify the migrate file. I don't think ORMs, especially those inspired by Django, should force developers to spend time in SQL code. Is there a reason why migration tool is one more years behind while Tortoise ORM is being actively developed and getting more attention?

**Comment by henadzit** [comment]
@licongy could you please list the main issues with Aerich that you encountered? If there is a prioritized list of issues, I can look at the most pressing issues.

Both tortoise and Aerich are open-source, so you can contribute too!

**Reply by licongy** [reply]
> @licongy could you please list the main issues with Aerich that you encountered? If there is a prioritized list of issues, I can look at the most pressing issues.
> 
> Both tortoise and Aerich are open-source, so you can contribute too!

Thank you for your reply. Firstly, I really like Tortoise ORM and now used it in our new project already. And I appreciate the efforts of the dev team.
One of the big bug with Aerich is that it looks like the upgrade and downgrade execution code is reversed when handling index changes to a model. When migrate performs a deletion of an index, it'll try to delete an index in upgrade part that didn't exist in the first place, and try to restore that index in downgrade. This happens in postgresql, I haven't tested if its the same issue in mysql and sqllite.
I will find time to look at the aerich code deep and try to contribute some pulls.

**Reply by henadzit** [reply, marked as answer]
It would be great if you could create a ticket in https://github.com/tortoise/aerich with a code sample. Maybe, there is already a ticket there since there are a lot of them. I'll try to have a go at this issue then.

**Reply by henadzit** [reply]
FYI, I'm working on [another schema migration](https://github.com/henadzit/tortoise-pathway) tool for Tortoise. It is in early development and by no means is ready for production but I think the foundation is right. 

Features:
- Generate schema migrations from Tortoise models
- Apply and revert migrations
- No need for a database connection to generate migrations
- If you have used Django migrations, you will feel at home
- SQLite and Postgres seems to work but expect a lot of bugs


A migration example: 

```python

class AutoMigration(Migration):
    """
    Auto-generated migration based on model changes.
    """

    dependencies = []
    operations = [
        CreateModel(
            model="models.Post",
            fields={
                "id": IntField(primary_key=True, unique=True),
                "title": CharField(max_length=255),
                "content": TextField(),
                "author": ForeignKeyFieldInstance(
                    "models.User",
                    related_name="posts",
                    on_delete="CASCADE",
                    source_field="author_id",
                ),
                "created_at": DatetimeField(auto_now_add=True),
                "updated_at": DatetimeField(auto_now_add=True),
            },
        ),
        CreateModel(
            model="models.User",
            fields={
                "id": IntField(primary_key=True, unique=True),
                "name": CharField(max_length=255),
                "email": CharField(unique=True, max_length=255),
            },
        ),
    ]
```

Code contributions and assistance with testing is welcome and appreciated!
